### ISSUE TITLE
onAuthenticationFailed should return plugin error

### ISSUE THREAD

**Issue by BorntraegerMarc** [NiklasMerz/cordova-plugin-fingerprint-aio#247]
## Description

This error handler is not implemented properly: https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/blob/7541dc3ebfaa1c3f7e59ea220ba51ba317d148bc/src/android/BiometricActivity.java#L81-L83

It doesn't call `onError`: https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/blob/7541dc3ebfaa1c3f7e59ea220ba51ba317d148bc/src/android/BiometricActivity.java#L117

This leads to a bug that the plugin never returns any error to cordova & thus nothing happens.

* What did you do?
On my OnePlus 6T device I executed the in-display fingerprint scanner & used the **wrong** finger to authenticate.
* What did happen?
Nothing happened. After debugging I found this bug.
* What did you expect to happen?
That the plugin returns an error code so the webview can inform the user.

## Environment

* Plugin version 3.0.1

```
Installed platforms:
  android 8.1.0
  ios 5.1.1
Available platforms: 
  browser ^6.0.0
  electron ^1.0.0
  osx ^5.0.0
  windows ^7.0.0
```

**I can provide a fix for this but have a question:**

Which error should be returned?

**Comment by usahunps** [comment]
Ah... I'm wondering since morning why is it not working on my OP6T. Me too looking for the solution of the same.

**Comment by BorntraegerMarc** [comment]
I can provide a fix for this bug but have a question to the experts: Which error should be returned?

**Comment by usahunps** [comment]
Any updates on this?

**Comment by BorntraegerMarc** [comment]
Nope. Still waiting on some guidance from any plugin authors (cc @NiklasMerz ) to know what error should be returned.

I'm happy to provide a PR

**Comment by usahunps** [comment]
@NiklasMerz, Can we get some help from you on this?

**Comment by NiklasMerz** [comment]
Please check out the branch from PR #236 . This PR contains major refactorings and may work differently. I am still trying to get this stable for the next version.

**Comment by Abhishek-Kanitkar** [comment]
@BorntraegerMarc I am using this plugin's 3.0.1 version.

Is the issue reported by you resolved as of now in any of the later versions ?

@NiklasMerz  kindly reply so that I can update to that version.

I am using Ionic 3 & I am also facing the same issue on One Plus 6T.

**Comment by NiklasMerz** [comment]
Please test the new version 4 and report issues.

Closing this. Please create new issues. Thank you.

**Comment by Abhishek-Kanitkar** [comment]
@NiklasMerz @BorntraegerMarc I upgraded to latest version of this plugin, i.e. 4.0.0. But I am still facing the same issue on One Plus 6T.

@NiklasMerz  what are your thoughts on the identification of possible solution by @BorntraegerMarc  ??

Can you please guide here ?

This issue is marked as closed but it hasn't been fixed yet.

@NiklasMerz . Please help.

**Comment by NiklasMerz** [comment]
@Abhishek-Kanitkar Please be careful with mentioning people. You don't have to use this everytime because people in this issue are already subscribed. We do our best to respond here.

I can now reproduce this on my OP 6T. The Biometric API is implemented strangely by OnePlus. If a enter an incorrect finger on a Samsung phone or the Emulator the native prompt shows me just that but OnePlus just shows nothing.

@Abhishek-Kanitkar It would be awesome if you could try out if the method Marc mentioned `onAuthenticationFailed` gets called in this case. If yes this probably should just call the onError method with the `BIOMETRIC_AUTHENTICATION_FAILED` error. If you could submit a pull request I might be able to review it soon and push a new version.

**Comment by Abhishek-Kanitkar** [comment]
@NiklasMerz  sorry to say but I don't have any experience with working on Native mobile app code. So if I have to debug & raise a PR for this issue, it will mean starting everything from scratch.

**Comment by Abhishek-Kanitkar** [comment]
> ### Bug report
> ## Description
> This error handler is not implemented properly:
> 
> https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/blob/7541dc3ebfaa1c3f7e59ea220ba51ba317d148bc/src/android/BiometricActivity.java#L81-L83
> 
> It doesn't call `onError`:
> 
> https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/blob/7541dc3ebfaa1c3f7e59ea220ba51ba317d148bc/src/android/BiometricActivity.java#L117
> 
> This leads to a bug that the plugin never returns any error to cordova & thus nothing happens.
> 
> * What did you do?
>   On my OnePlus 6T device I executed the in-display fingerprint scanner & used the **wrong** finger to authenticate.
> * What did happen?
>   Nothing happened. After debugging I found this bug.
> * What did you expect to happen?
>   That the plugin returns an error code so the webview can inform the user.
> 
> ## Environment
> * Plugin version 3.0.1
> 
> ```
> Installed platforms:
>   android 8.1.0
>   ios 5.1.1
> Available platforms: 
>   browser ^6.0.0
>   electron ^1.0.0
>   osx ^5.0.0
>   windows ^7.0.0
> ```
> 
> **I can provide a fix for this but have a question:**
> 
> Which error should be returned?



> @Abhishek-Kanitkar Please be careful with mentioning people. You don't have to use this everytime because people in this issue are already subscribed. We do our best to respond here.
> 
> I can now reproduce this on my OP 6T. The Biometric API is implemented strangely by OnePlus. If a enter an incorrect finger on a Samsung phone or the Emulator the native prompt shows me just that but OnePlus just shows nothing.
> 
> @Abhishek-Kanitkar It would be awesome if you could try out if the method Marc mentioned `onAuthenticationFailed` gets called in this case. If yes this probably should just call the onError method with the `BIOMETRIC_AUTHENTICATION_FAILED` error. If you could submit a pull request I might be able to review it soon and push a new version.



> ### Bug report
> ## Description
> This error handler is not implemented properly:
> 
> https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/blob/7541dc3ebfaa1c3f7e59ea220ba51ba317d148bc/src/android/BiometricActivity.java#L81-L83
> 
> It doesn't call `onError`:
> 
> https://github.com/NiklasMerz/cordova-plugin-fingerprint-aio/blob/7541dc3ebfaa1c3f7e59ea220ba51ba317d148bc/src/android/BiometricActivity.java#L117
> 
> This leads to a bug that the plugin never returns any error to cordova & thus nothing happens.
> 
> * What did you do?
>   On my OnePlus 6T device I executed the in-display fingerprint scanner & used the **wrong** finger to authenticate.
> * What did happen?
>   Nothing happened. After debugging I found this bug.
> * What did you expect to happen?
>   That the plugin returns an error code so the webview can inform the user.
> 
> ## Environment
> * Plugin version 3.0.1
> 
> ```
> Installed platforms:
>   android 8.1.0
>   ios 5.1.1
> Available platforms: 
>   browser ^6.0.0
>   electron ^1.0.0
>   osx ^5.0.0
>   windows ^7.0.0
> ```
> 
> **I can provide a fix for this but have a question:**
> 
> Which error should be returned?

@BorntraegerMarc  as per above comment by Niklas, if onAuthenticationFailed gets called in your case, then can you please call the onError method with the BIOMETRIC_AUTHENTICATION_FAILED error inside of onAuthenticationFailed.

Please raise a PR if you find that the above changes have fixed this issue.

Thanks in advance !! I will also try to learn how to setup a Cordova project & make changes in it and run it as well.

**Comment by NiklasMerz** [comment]
No worries. It's not that hard if you want to invest some time in trying it out. You could start by going into your native project in the platforms/android folder opening it with Android studio and trying to debug with `System.out.println`.

I maybe can look into that soon. Keep in mind I am just doing that in my free time.

**Comment by Abhishek-Kanitkar** [comment]
@NiklasMerz  I don't have OnePlus 6T device handy with me like you, to do USB debugging from my laptop. My colleague has it & with the lockdown, it's not possible to meet her in person. So I am not sure how System.out.println will help. I would have loved to debug this issue but I have constraints in terms of native technology expertise. Thats why I request you or Marc to look into it.

For my project, this is a serious issue because it blocks the user from getting authenticated in any other way as we don't get any response from the plugin in error scenario.

Whenever you or Marc get sometime from your busy schedule, kindly provide a fix for this issue. The sooner the better for us obviously :)

Thanks in advance again & please don't misunderstand me !

**Comment by NiklasMerz** [comment]
@Abhishek-Kanitkar I understand! That's totally fine. Luckily a OP 6T is my personal and only device I have 😄 

In my tests I could hit the back button to cancel the authentication after a failed attempt. Then you will get the cancelled response.

I will see if I can have a look at this in the next few days.

**Comment by Abhishek-Kanitkar** [comment]
Thanks @NiklasMerz  ! Yes I have observed that too, that if you press back button then you get response but thats not intuitive to the user ;( . Anyways, thanks a lot & I will be looking forward to the fix for this  :)

By any chance, do you have any clue why Face detection on OnePlus 6T devices is not recognized by this plugin ?

Is there something that you can do about it ? It's an issue with OnePlus devices in general.

**Comment by NiklasMerz** [comment]
It has taken a long time but thank you all for persisting through this issue. I just pushed an update to NPM with a fix for this.

> By any chance, do you have any clue why Face detection on OnePlus 6T devices is not recognized by this plugin ?

I think this is a similar restriction of OnePlus' implementation for Biometric prompt. Not sure we can do something about it. Please open a new issue for discussing this if necessary.

**Comment by Abhishek-Kanitkar** [comment]
@NiklasMerz  Thanks a lot for fixing this issue !!! Do you think the Ionic wrapper for this plugin, gets updated with latest version automatically ?

https://ionicframework.com/docs/v3/native/fingerprint-aio/

**Comment by NiklasMerz** [comment]
Ionic Native maybe needs to get updated. Your link is for an old version. This is the current documentation: https://ionicframework.com/docs/native/fingerprint-aio

In the latest version 5.29.0 there are already variables and functions for the new feature: https://github.com/ionic-team/ionic-native/blob/master/src/@ionic-native/plugins/fingerprint-aio/index.ts

Please test it. You can fix this yourself if you create an pull request for the ionic native repo. Otherwise I can help with that too.

**Comment by Abhishek-Kanitkar** [comment]
@NiklasMerz  I have two questions.
1) Does the new version of this plugin expect you to have minimum cordova-android version as 9.0.0 ?

2) In my project, we can't use latest version of Ionic in near future. We have to work with Ionic V3. What are the steps to ensure that the Ionic wrapper dowloads the latest version of this plugin ?

Currently I could do so by changing the version of this plugin in package.json of my Ionic project. 

Can you please throw some light on above 2 points ?

**Comment by BorntraegerMarc** [comment]
Thx for fixing this 🙂 I was on vacation so didn't see these messages. We'll test it out asap

> Thanks a lot for fixing this issue !!! Do you think the Ionic wrapper for this plugin, gets updated with latest version automatically ?

@Abhishek-Kanitkar for this fix you can update the cordova version without ionic native. ionic native is just for typing (and dependency injection). It does not directly depend on the native cordova plugin code.

**Comment by Abhishek-Kanitkar** [comment]
> Thx for fixing this 🙂 I was on vacation so didn't see these messages. We'll test it out asap
> 
> > Thanks a lot for fixing this issue !!! Do you think the Ionic wrapper for this plugin, gets updated with latest version automatically ?
> 
> @Abhishek-Kanitkar for this fix you can update the cordova version without ionic native. ionic native is just for typing (and dependency injection). It does not directly depend on the native cordova plugin code.

So @BorntraegerMarc if I have already used Ionic wrapper for this plugin in my code, then how can I upgrade the cordova plugin version that the wrapper is pointing to ? By changing it in my Ionic project's package.json file ?

Can you please suggest how I can still keep using Ionic wrapper ? Or are you saying that I should only use the cordova plugin as it is directly ?

**Comment by BorntraegerMarc** [comment]
@Abhishek-Kanitkar find `cordova-plugin-fingerprint-aio` in your `package.json` and change the version to `v4.0.1`. After that run `npm install`

Voilà: You have now the latest version of this plugin

**Comment by Abhishek-Kanitkar** [comment]
> @Abhishek-Kanitkar find `cordova-plugin-fingerprint-aio` in your `package.json` and change the version to `v4.0.1`. After that run `npm install`
> 
> Voilà: You have now the latest version of this plugin

Thanks ! So the Ionic wrapper plugin will start using the latest cordova plugin version right ?
But what if the cordova plugin api has changed but Ionic wrapper plugin hasn't updated it's definition ?

**Comment by Abhishek-Kanitkar** [comment]
@NiklasMerz  sorry to disturb you. Can you please check once on your OnePlus 6T device with latest build @4.0.1 ?

Seems like the fix given is not working any more  :(

Can you please confirm if it's working on your phone ?

**Comment by Abhishek-Kanitkar** [comment]
> Thx for fixing this 🙂 I was on vacation so didn't see these messages. We'll test it out asap
> 
> > Thanks a lot for fixing this issue !!! Do you think the Ionic wrapper for this plugin, gets updated with latest version automatically ?
> 
> @Abhishek-Kanitkar for this fix you can update the cordova version without ionic native. ionic native is just for typing (and dependency injection). It does not directly depend on the native cordova plugin code.

Hi @BorntraegerMarc 

Can you please try to validate now, whether the changes done to fix the issue raised by you here, are still working on your OnePlus 6T device or not ?

When I tried yesterday, it didn't return anything in error callback. It wasn't invoked.

**Comment by BorntraegerMarc** [comment]
Sorry, I don’t have a oneplus any longer....

**Comment by Abhishek-Kanitkar** [comment]
> Sorry, I don’t have a oneplus any longer....

Ohh

**Comment by Abhishek-Kanitkar** [comment]
@NiklasMerz  can you please verify on your OnePlus device whether the fix provided you in @4.0.1 is still working or not ?

Please respond.

**Comment by NiklasMerz** [comment]
@Abhishek-Kanitkar Please stop spamming this issue. I  cannot give respong immediatly on closed issues and support questions. I have would lock closed issues to stop this, but I try to avoid that most of the time. 

I work on this plugin in my free time and when I like to. Many people are using it and counting on my support but I cannot provide that for free. I just don't have the time and motivation to do that. I would love to see more people using Github sponsors more if they are happy with this plugin.

For support questions I just enabled Github discussions.
